File: CyxbsMobile2019_iOS/Login/Model/LoginModel.m
//
//  LoginModel.m
//  CyxbsMobile2019_iOS
//
//  Created by 方昱恒 on 2019/10/23.
//  Copyright © 2019 Redrock. All rights reserved.
//

#import "LoginModel.h"
#import "UserItem.h"


@implementation LoginModel
//
//- (void)loginWithStuNum:(NSString *)stuNum
//               andIdNum:(NSString *)idNum
//              succeeded:(void (^)(NSDictionary * _Nonnull))succeeded
//                 failed:(void (^)(NSError * _Nonnull))failed {
//    NSDictionary *params = @{
//        @"stuNum": stuNum,
//        @"idNum": idNum
//    };
//
//    AFHTTPSessionManager *manager = [AFHTTPSessionManager manager];
//    // 这个请求需要上传json
//    manager.requestSerializer = [AFJSONRequestSerializer serializer];
//
//    [manager POST:LOGINAPI parameters:params success:^(NSURLSessionDataTask * _Nonnull task, id  _Nonnull responseObject) {
//        // 返回的responseObject[@"data"][@"token"]是一个base64编码的字符串
//        // 以'.'分隔为前后两段。
//        // 前半段base64解码为用户信息的json字符串，后面半段为签名。
//        NSString *token = responseObject[@"data"][@"token"];
//        NSString *payload_BASE64 = [token componentsSeparatedByString:@"."][0];
//
//        // 这里要把json字符串转换为字典
//        NSData *payloadData = [[NSData alloc] initWithBase64EncodedString:payload_BASE64 options:0];
//        NSError *error;
//        NSMutableDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:payloadData options:NSJSONReadingMutableContainers error:&error];
//        jsonObject[@"token"] = responseObject[@"data"][@"token"];
//        jsonObject[@"refreshToken"] = responseObject[@"data"][@"refreshToken"];
//
//        // 执行回调
//        succeeded(jsonObject);
//        if (error) {
//            failed(error);
//        }
//    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
//        failed(error);
//    }];
//}

- (void)loginWithStuNum:(NSString *)stuNum
               andIdNum:(NSString *)idNum
              succeeded:(void (^)(NSDictionary * _Nonnull))succeeded
                 failed:(void (^)(NSError * _Nonnull))failed {
    NSDictionary *parameters = @{
                                 @"stuNum": stuNum,
                                 @"idNum": idNum
                                 };
    
    HttpClient *client = [HttpClient defaultClient];
    [client requestWithJson:LOGINAPI method:HttpRequestPost parameters:parameters prepareExecute:nil progress:nil success:^(NSURLSessionDataTask *task, id responseObject) {
        
        NSString *token = responseObject[@"data"][@"token"];
        NSString *payload_BASE64 = [token componentsSeparatedByString:@"."][0];
        
        // 这里要把json字符串转换为字典
        NSData *payloadData = [[NSData alloc] initWithBase64EncodedString:payload_BASE64 options:0];
        NSError *error;
        NSMutableDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:payloadData options:NSJSONReadingMutableContainers error:&error];
        jsonObject[@"token"] = responseObject[@"data"][@"token"];
        jsonObject[@"refreshToken"] = responseObject[@"data"][@"refreshToken"];
        
        // 执行回调
        succeeded(jsonObject);
        if (error) {
            failed(error);
        }
    } failure:^(NSURLSessionDataTask *task, NSError *error) {
        failed(error);
    }];
}

@end

File: CyxbsMobile2019_iOS/Swift/Login/Model/LoginModel.swift
//
//  LoginModel.swift
//  CyxbsMobile2019_iOS
//
//  Created by 方昱恒 on 2020/10/5.
//  Copyright © 2020 Redrock. All rights reserved.
//

import UIKit
import SwiftyJSON

class LoginModel: NSObject {
    
    class func loginWith(stuNum: String,
                         idNum: String,
                         success: @escaping () -> Void,
                         failed: @escaping () -> Void) {
        let params = [
            "stuNum": stuNum,
            "idNum": idNum
        ]

        let LoginURL = BaseURL() as String + "magipoke/token"
        guard let url = URL(string:LoginURL) else { return }
        
        SwiftyClient.shared.request(url, method: .post, headers: nil, parameters: params) { (response) in
            
            guard let response = response else {
                failed()
                return
            }
            
            let json = JSON(response)
            
            if let token = json["data"]["token"].string,
               let refreshToken = json["data"]["refreshToken"].string {
                let item = UserItemTool.defaultItem();
                // 保存token、refreshToke、账号、密码
                item.token = token
                item.refreshToken = refreshToken
                item.firstLogin = true
                UserDefaultTool.saveStuNum(stuNum)
                UserDefaultTool.saveIdNum(idNum)
                
                // 解码用户信息并保存
                let userInfo_Base64 = token.split(separator: ".")[0]
                if let data = Data(base64Encoded: String(userInfo_Base64)) {
                    do {
                        let userInfo = try JSON(data: data)
                        UserItem.mj_object(withKeyValues: userInfo.dictionaryObject ?? nil)
                        
                        success()
                        //
                        UserDefaults.standard.setValue(getLastLogInTime(), forKey: LastLogInTimeKey_double)
                        NotificationCenter.default.postNotificationOnMainThread(withName: "Login_LoginSuceeded", object: nil, userInfo: ["userItem": item])
                        
                        
                        
                    } catch {
                        // 解码失败调用failed
                        failed()
                    }
                }
                
            } else {
                failed()
            }
            
            
        }
        
    }
    class func getLastLogInTime() -> TimeInterval {
        return Date.timeIntervalSinceReferenceDate + Date.timeIntervalBetween1970AndReferenceDate
    }
    
    class func BaseURL() -> NSString {
        let baseURL = UserDefaults.standard.string(forKey: "baseURL")
        return baseURL! as NSString
    }

}
